pkg/api/vmtemplate/store.go

```diff
@@ -0,0 +1,138 @@
+package vmtemplate
+
+import (
+	"sync"
+
+	"github.com/pkg/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+
+	"github.com/rancher/apiserver/pkg/apierror"
+	"github.com/rancher/apiserver/pkg/types"
+	ctlvmv1alpha1 "github.com/rancher/harvester/pkg/generated/controllers/vm.cattle.io/v1alpha1"
+	"github.com/rancher/harvester/pkg/ref"
+	"github.com/rancher/wrangler/pkg/schemas/validation"
+)
+
+const (
+	templateLabel = "template.vm.cattle.io/templateID"
+)
+
+type templateVersionStore struct {
+	types.Store
+	templateCache        ctlvmv1alpha1.TemplateCache
+	templateVersions     ctlvmv1alpha1.TemplateVersionClient
+	templateVersionCache ctlvmv1alpha1.TemplateVersionCache
+	mu                   sync.Mutex
+}
+
+type templateStore struct {
+	types.Store
+	templateVersionCache ctlvmv1alpha1.TemplateVersionCache
+}
+
+func (s *templateVersionStore) Create(request *types.APIRequest, schema *types.APISchema, data types.APIObject) (types.APIObject, error) {
+	ns := data.Data().String("metadata", "namespace")
+	templateID := data.Data().String("spec", "templateId")
+	if templateID == "" {
+		return types.APIObject{}, apierror.NewAPIError(validation.InvalidBodyContent, "TemplateId is empty")
+	}
+
+	templateNs, templateName := ref.Parse(templateID)
+	if ns != templateNs {
+		return types.APIObject{}, apierror.NewAPIError(validation.InvalidBodyContent, "Template version and template should belong to same namespace")
+	}
+
+	if err := s.setOwnerRefAndLabel(request, data); err != nil {
+		return types.APIObject{}, apierror.NewAPIError(validation.InvalidBodyContent, err.Error())
+	}
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	sets := labels.Set{
+		templateLabel: templateName,
+	}
+	existedVersions, err := s.templateVersions.List(ns, metav1.ListOptions{LabelSelector: sets.AsSelector().String()})
+	if err != nil {
+		return types.APIObject{}, apierror.NewAPIError(validation.ServerError, err.Error())
+	}
+
+	currentVersion := len(existedVersions.Items) + 1
```

Comment:
either set the `versionNumber` to the type of `int` or must convert value this to a string.

Reply:
updated